File: algorithms/173. Binary Search Tree Iterator/solution.py
# Definition for a  binary tree node
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator(object):
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.path = []
        cur = root
        while cur:
            self.path.append(cur)
            cur = cur.left

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.path

    def next(self):
        """
        :rtype: int
        """
        node = self.path.pop()
        cur = node.right
        while cur:
            self.path.append(cur)
            cur = cur.left
        return node.val

# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())

File: algorithms/173. Binary Search Tree Iterator/solution.cc
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
    stack<TreeNode *> s;
public:
    BSTIterator(TreeNode *root) {
        if ( root ) {
            s.push(root);
            while ( s.top()->left ) {
                s.push(s.top()->left);
            }
        }
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return s.size() > 0;
    }

    /** @return the next smallest number */
    int next() {
        TreeNode * node = s.top();
        s.pop();
        if ( node->right ) {
            s.push(node->right);
            while ( s.top()->left ) {
                s.push(s.top()->left);
            }
        }
        return node->val;
    }
};

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */
